Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,191 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A cache that enables the performant persistence of individual messages to disk.
+/// This cache is intended to be written from and appended to from the same serial queue.
+public final class CacheAdvance<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - file: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache leads to undefined behavior.
+    public init(
+        file: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
+        throws
+    {
+        self.file = file
+        self.maximumBytes = maximumBytes
+        self.shouldOverwriteOldMessages = shouldOverwriteOldMessages
+
+        writer = try FileHandle(forWritingTo: file)
+        reader = try FileHandle(forReadingFrom: file)
+
+        lengthOfMessageSuffix = {
+            if shouldOverwriteOldMessages {
+                /// The message suffix requires space for both `endOfNewestMessageMarker` and `offsetInFileOfOldestMessage` after each message when rolling is enabled.
+                return Bytes(Data.messageSpanLength + Data.oldestMessageOffsetLength)
+            } else {
+                /// The message suffix requires space for `endOfNewestMessageMarker` after each message.
+                return Bytes(Data.messageSpanLength)
+            }
+        }()
+    }
+
+    deinit {
+        try? writer.close()
+        try? reader.close()
+    }
+
+    // MARK: Public
+
+    public let file: URL
+
+    /// Appends a message to the cache.
+    /// - Parameter message: A message to write to disk.
+    public func append(message: T) throws {
+        try setUpFileHandlesIfNecessary()
+
+        let encodedMessage = EncodableMessage(message: message, encoder: encoder)
+        let messageData = try encodedMessage.encodedData()
+        let cacheHasSpaceForNewMessage = writer.offsetInFile + Bytes(messageData.count) + lengthOfMessageSuffix <= maximumBytes
```

Comment:
What do you think about making this two lines or making it a property or method?

Reply:
I'll break `Bytes(messageData.count) + lengthOfMessageSuffix` into it's own method, since I use that in multiple spots.